FWIO/RNTuple/plugins/RNTupleOutputFile.cc

```diff
@@ -0,0 +1,583 @@
+          if (index >= iUseStreamer.size() or not iUseStreamer[index]) {
+            auto field = ROOT::Experimental::RFieldBase::Create(fixBranchName(prod.first->branchName()),
+                                                                prod.first->wrappedName())
+                             .Unwrap();
+            if (noSplitSubFields) {
+              //use the 'conventional' way to store fields
+              for (auto& subfield : *field) {
+                noSplitField(subfield);
+              }
+            } else if (not iNoSplitFields.empty()) {
+              checkSubFieldsForNoSplit(*field, iNoSplitFields);
+            }
+            iModel.AddField(std::move(field));
+            branchesWithStoredHistory_.insert(prod.first->branchID());
+          } else {
+            auto field = std::make_unique<ROOT::Experimental::RStreamerField>(fixBranchName(prod.first->branchName()),
+                                                                              prod.first->wrappedName());
+            iModel.AddField(std::move(field));
+            branchesWithStoredHistory_.insert(prod.first->branchID());
+          }
+        });
+        ++index;
+      } catch (cms::Exception& iExcept) {
+        std::ostringstream s;
+        s << "while setting up field " << prod.first->branchName();
+        iExcept.addContext(s.str());
+        throw;
+      }
+    }
+  }
+
+  std::vector<RNTupleOutputFile::Product> RNTupleOutputFile::associateDataProducts(SelectedProducts const& iProducts,
+                                                                                   RNTupleModel const& iModel) {
+    std::vector<Product> ret;
+    ret.reserve(iProducts.size());
+    for (auto const& prod : iProducts) {
+      ret.emplace_back(prod.second, prod.first, iModel.GetToken(fixBranchName(prod.first->branchName())));
+    }
+    return ret;
+  }
+
+  void RNTupleOutputFile::setupRuns(SelectedProducts const& iProducts, Config const& iConfig) {
+    std::string kRunAuxName = "RunAuxiliary";
+    {
+      auto model = ROOT::Experimental::RNTupleModel::CreateBare();
+      {
+        auto field = ROOT::Experimental::RFieldBase::Create(kRunAuxName, "edm::RunAuxiliary").Unwrap();
+        model->AddField(std::move(field));
+      }
+      const std::vector<bool> streamerNothing;
+      const std::vector<std::string> unsplitNothing;
+      setupDataProducts(iProducts, streamerNothing, unsplitNothing, *model);
+
+      auto writeOptions = ROOT::Experimental::RNTupleWriteOptions();
+      writeOptions.SetCompression(convert(iConfig.compressionAlgo), iConfig.compressionLevel);
+      runs_ = ROOT::Experimental::RNTupleWriter::Append(std::move(model), "Runs", file_, writeOptions);
+    }
+    products_[InRun] = associateDataProducts(iProducts, runs_->GetModel());
+    runAuxField_ = runs_->GetModel().GetToken(kRunAuxName);
+  }
+  void RNTupleOutputFile::setupLumis(SelectedProducts const& iProducts, Config const& iConfig) {
```

Comment:
The code in `setupRuns()` and `setupLumis()` seems nearly identical. Would it make sense to abstract the common parts?